### DISCUSSION TITLE
To fork or not to fork

### DISCUSSION THREAD

**Discussion by digitarald** [yandeu/five-server#4] [General]
Hi there. I am from the VS Code team and saw your comments [here](https://github.com/microsoft/vscode-brackets-pack/pull/1). First off, kudos! I love the project and the quality of life features it already added. I wanted to learn more about the intent of Five Server vs Live Server, given that this project is meant to be a fork (but not using the GitHub forking flow).

There are trade-offs for forking into a new extension. VS Code does not offer paths to migrate users off extensions; so users get stuck on unmaintained extensions. As _ritwickdey_, the creator of Live Server, is looking for maintainers (*"I'm sorry but I'm super busy now. If you want to be a maintainer of the project, please feel free to contact me!…"*) – was it considered to invest into the existing extension?

Thanks for helping me understand!

**Comment by yandeu** [comment]
Hi @digitarald, glad you joined!

> was it considered to invest into the existing extension

No. This is not an option.

## The Story

[vscode-live-server (vscode extension)](https://github.com/ritwickdey/vscode-live-server) is based on the [live-server (npm)](https://www.npmjs.com/package/live-server). Both are not updated in years. Many dependencies and sub-dependencies of live-server are out-of-date. Users of macOS Big Sur recently reported many issues regarding the [opn package](https://www.npmjs.com/package/opn) live-server uses. In order to update all dependencies and remove potential vulnerabilities, I had to fork live-server. So, I forked live-server, updated all dependencies, rewrote everything in TypeScript, and even forked some outdated sub-dependencies directly into the source code. The project is available under the name [five-server (npm)](https://www.npmjs.com/package/five-server). five-server is now up-to-date and very easy to maintain. Actually, five-server and five-server-vscode do not include one single line from vscode-live-server. The only thing in common is the resembling logo.

## Future

There are many great features yet to come. I have just (nearly) finished a sort of Hot Module Replacement. Means, updating the `<body>` in real-time without touching the current state (modified by JavaScript) of the DOM.

More Soon:
- Real-Time HTML Validation (partially implemented; released soon)
- Highlight for CSS
- Browser Sync
- and more...

**Reply by digitarald** [reply]
Thanks for the story and insights into your future plans!

> No. This is not an option.

Even if _Five Server_ is a complete rewrite, could it still replace the current Live Server extension as a new major version increment? From what you mentioned I only see implementation changes that should not block migration, or am I missing something?

**Reply by yandeu** [reply]
Yes, with minor exceptions:

- I have changed the default port to 5555.
- ~~I have not implemented any shortcuts yet.~~
- ~~Five Server does not use `contributes.configuration` yet. It uses `fiveserver.config.js` for now.~~
- ~~Live Server also supports SVG files. I have not added support for SVG yet.~~
- Live Server writes "Multi-root workspace supported". I'm not sure what Ritwick means by that. _Do you?_
- Live Server features a [web extension](https://github.com/ritwickdey/live-server-web-extension). I have built one, but not published yet.
- ~~I have removed htpasswd support. (not sure if it is included in "Live Server VSCode")~~
- ~~I have removed support for custom https modules. (not sure if it is included in "Live Server VSCode")~~
- _Maybe some other minor things I forgot to mention._

**Reply by lukas-kasticky** [reply]
> Live Server writes "Multi-root workspace supported". I'm not sure what Ritwick means by that. Do you?

This is referring to a [VS Code feature](https://code.visualstudio.com/docs/editor/workspaces#_multiroot-workspaces). You can create workspaces with folders in different directories. Five Server doesn't support this unfortunately.
